File: app/src/main/java/vn/loitp/app/activity/demo/ebookwithrealm/realm/RealmController.java
package vn.loitp.app.activity.demo.ebookwithrealm.realm;

import android.app.Activity;
import android.app.Application;

import androidx.fragment.app.Fragment;

import io.realm.Realm;
import io.realm.RealmResults;
import vn.loitp.app.activity.demo.ebookwithrealm.model.Book;

public class RealmController {

    private static RealmController instance;
    private final Realm realm;

    private RealmController(Application application) {
        realm = Realm.getDefaultInstance();
    }

    public static RealmController with(final Fragment fragment) {
        if (instance == null) {
            instance = new RealmController(fragment.getActivity().getApplication());
        }
        return instance;
    }

    public static RealmController with(final Activity activity) {
        if (instance == null) {
            instance = new RealmController(activity.getApplication());
        }
        return instance;
    }

    public static RealmController with(final Application application) {
        if (instance == null) {
            instance = new RealmController(application);
        }
        return instance;
    }

    public static RealmController getInstance() {
        return instance;
    }

    public Realm getRealm() {
        return realm;
    }

    //Refresh the realm istance
    public void refresh() {
        realm.refresh();
    }

    //clear all objects from Book.class
    public void clearAll() {
        realm.beginTransaction();
        realm.clear(Book.class);
        realm.commitTransaction();
    }

    //find all objects in the Book.class
    public RealmResults<Book> getBooks() {
        return realm.where(Book.class).findAll();
    }

    //query a single item with the given id
    public Book getBook(String id) {
        return realm.where(Book.class).equalTo("id", id).findFirst();
    }

    //check if Book.class is empty
    public boolean hasBooks() {
        return !realm.allObjects(Book.class).isEmpty();
    }

    //query example
    public RealmResults<Book> queryedBooks() {
        return realm.where(Book.class)
                .contains("author", "Author 0")
                .or()
                .contains("title", "Realm")
                .findAll();

    }
}

File: app/src/main/java/vn/loitp/app/activity/demo/ebookWithRealm/realm/RealmController.kt
package vn.loitp.app.activity.demo.ebookWithRealm.realm

import android.app.Activity
import android.app.Application
import androidx.fragment.app.Fragment
import io.realm.Realm
import io.realm.RealmResults
import vn.loitp.app.activity.demo.ebookWithRealm.model.Book

class RealmController private constructor() {
    val realm: Realm = Realm.getDefaultInstance()

    // Refresh the realm istance
    fun refresh() {
        realm.refresh()
    }

    // clear all objects from Book.class
    @Suppress("unused")
    fun clearAll() {
        realm.beginTransaction()
        realm.clear(Book::class.java)
        realm.commitTransaction()
    }

    // find all objects in the Book.class
    val books: RealmResults<Book>
        get() = realm.where(Book::class.java).findAll()

    // query a single item with the given id
    @Suppress("unused")
    fun getBook(id: String?): Book {
        return realm.where(Book::class.java).equalTo("id", id).findFirst()
    }

    // check if Book.class is empty
    @Suppress("unused")
    fun hasBooks(): Boolean {
        return !realm.allObjects(Book::class.java).isEmpty()
    }

    // query example
    @Suppress("unused")
    fun queryedBooks(): RealmResults<Book> {
        return realm.where(Book::class.java)
            .contains("author", "Author 0")
            .or()
            .contains("title", "Realm")
            .findAll()
    }

    companion object {
        @JvmStatic
        var instance: RealmController? = null
            private set

        fun with(fragment: Fragment): RealmController? {
            if (instance == null) {
                instance = RealmController()
            }
            return instance
        }

        fun with(activity: Activity): RealmController? {
            if (instance == null) {
                instance = RealmController()
            }
            return instance
        }

        fun with(application: Application): RealmController? {
            if (instance == null) {
                instance = RealmController()
            }
            return instance
        }
    }
}
